### PR TITLE
Implement blockcipher module for layer data encrypt/decrypt

### PR DESCRIPTION
**Pull request by arronwy**

### CODE REVIEW

File: src/blockcipher/mod.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct PrivateLayerBlockCipherOptions {
+    /// symmetric_key represents the symmetric key used for encryption/decryption
+    /// This field should be populated by encrypt/decrypt calls
+    #[serde(rename = "symkey")]
+    pub symmetric_key: Vec<u8>,
+
+    /// digest is the digest of the original data for verification.
+    /// This is NOT populated by encrypt/decrypt calls
+    pub digest: String,
```
Comment (pravinrajr9) [inline src/blockcipher/mod.rs:29]: Shouldn't we add the serde rename explicitly for all fields?

Comment (arronwy) [reply src/blockcipher/mod.rs:29]: Rust serialize and deserialize will use the parameter/rust name by default, we only need rename it when we have different name with the golang version.

File: src/blockcipher/aes_ctr.rs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            }
+        }
+
+        let cipher = Aes256Ctr::new(
+            GenericArray::from_slice(symmetric_key.as_slice()),
+            GenericArray::from_slice(nonce.as_slice()),
+        );
+
+        let hmac =
+            HmacSha256::new_from_slice(symmetric_key.as_slice()).expect("hmac use symmetric key");
```
Comment (pravinrajr9) [inline src/blockcipher/aes_ctr.rs:145]: Should'nt we handle error here using match instead of expect, expect would result in panic right?

Comment (arronwy) [reply src/blockcipher/aes_ctr.rs:145]: hmac can take key of any size: https://docs.rs/hmac/0.11.0/hmac/ and will not panic , this expect is just compatiable with NewMac trait.

Comment (pravinrajr9) [reply src/blockcipher/aes_ctr.rs:145]: ok

**Comment by pravinrajr9** [comment]
LGTM

File: src/blockcipher/aes_ctr.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            return Ok(0);
+        }
+
+        let read_len = self.reader.as_mut().unwrap().read(buf)?;
+        if read_len == 0 {
+            self.done_encrypting = true;
+        }
+
+        if !self.encrypt {
+            self.hmac.as_mut().unwrap().update(&buf[0..read_len]);
```
Comment (sameo) [inline src/blockcipher/aes_ctr.rs:42]: Could we please avoid unwraping and do something like that instead:

```Rust
if !self.encrypt {
   if let Some(hmac) = self.hmac.as_mut() {
       hmac.update(&buf[0..read_len]);
   }
```

File: src/blockcipher/aes_ctr.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+            }
+        }
+
+        self.cipher
+            .as_mut()
+            .unwrap()
+            .apply_keystream(&mut buf[0..read_len]);
+
+        if self.encrypt {
+            self.hmac.as_mut().unwrap().update(&buf[0..read_len]);
```
Comment (sameo) [inline src/blockcipher/aes_ctr.rs:71]: Ditto

File: src/blockcipher/aes_ctr.rs (hunk 2 above)
Comment (sameo) [reply src/blockcipher/aes_ctr.rs:145]: @arronwy You're right, the Hmac implementation of `NewMac` does not panic. But this is an internal implementation and we should not rely on that. I agree with @pravinrajr9 we should handle the `Result` from `new_from_slice` and don't make any assumption on the internal implementation not panicking.

**Review by sameo — COMMENTED** [review]
The PR looks pretty good, I only have a few minor comments. Thanks @arronwy

File: src/blockcipher/aes_ctr.rs (hunk 3 above)
Comment (arronwy) [reply src/blockcipher/aes_ctr.rs:42]: Thanks, I'll fix it.

File: src/blockcipher/aes_ctr.rs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                hmac.update(&buf[0..read_len]);
+            }
+
+            // If we done encrypting, let the HMAC comparison
+            // provide a verdict
+            if self.done_encrypting
+                && self
+                    .hmac
+                    .as_ref()
+                    .unwrap()
```
Comment (sameo) [inline src/blockcipher/aes_ctr.rs:52]: Sorry if my initial review was not clear enough, but we should also remove this and the other unwrap() from this method, and gracefully handle the Option. Something like:

 ```Rust
if let Some(hmac) = self.hmac.as_ref() {
    hmac.clone().verify.map_err()?;
}

would work?

Comment (arronwy) [reply src/blockcipher/aes_ctr.rs:52]: Yes, it works, I'll fix all unwrap() operations

File: src/blockcipher/aes_ctr.rs (hunk 2 above)
Comment (arronwy) [reply src/blockcipher/aes_ctr.rs]: After remove the Err wraper, it will works.

**Review by sameo — APPROVED** [review]

**Pull request merged**